pkg/apis/intevents/v1alpha1/brokercell_defaults.go

```diff
@@ -35,22 +63,67 @@ func (bc *BrokerCell) SetDefaults(ctx context.Context) {
 
 // SetDefaults sets the default field values for a BrokerCellSpec.
 func (bcs *BrokerCellSpec) SetDefaults(ctx context.Context) {
-	if bcs.Components.Fanout.MinReplicas == nil {
-		bcs.Components.Fanout.MinReplicas = ptr.Int32(minReplicas)
+	// Fanout defaults
+	bcs.Components.Fanout.SetCPUDefaults(cpuRequestFanout, cpuLimitFanout)
+	bcs.Components.Fanout.SetMemoryDefaults(memoryLimitToRequestCoefficientFanout, memoryRequestFanout)
+	bcs.Components.Fanout.SetAutoScalingDefaults(targetMemoryUsageCoefficientFanout, avgCPUUtilizationFanout)
+	// Retry defaults
+	bcs.Components.Retry.SetCPUDefaults(cpuRequestRetry, cpuLimitRetry)
+	bcs.Components.Retry.SetMemoryDefaults(memoryLimitToRequestCoefficientRetry, memoryRequestRetry)
+	bcs.Components.Retry.SetAutoScalingDefaults(targetMemoryUsageCoefficientRetry, avgCPUUtilizationRetry)
+	// Ingress defaults
+	bcs.Components.Ingress.SetCPUDefaults(cpuRequestIngress, cpuLimitIngress)
+	bcs.Components.Ingress.SetMemoryDefaults(memoryLimitToRequestCoefficientIngress, memoryRequestIngress)
+	bcs.Components.Ingress.SetAutoScalingDefaults(targetMemoryUsageCoefficientIngress, avgCPUUtilizationIngress)
+}
+
+// SetMemoryDefaults sets the memory consumption related default field values for ComponentParameters.
+func (componentParams *ComponentParameters) SetMemoryDefaults(memoryLimitToRequestCoefficient float64, memoryRequest string) {
+	if componentParams.MemoryRequest == nil {
+		componentParams.MemoryRequest = ptr.String(memoryRequest)
+	}
+	requestedMemoryQuantity := resource.MustParse(*componentParams.MemoryRequest)
+	if componentParams.MemoryLimit == nil {
+		autoSelectedLimit := quantityutil.MultiplyQuantity(requestedMemoryQuantity, memoryLimitToRequestCoefficient)
+		componentParams.MemoryLimit = ptr.String(autoSelectedLimit.String())
+	}
+	// Make sure the limit is not lower than what's requested
+	memoryLimitQuantity := resource.MustParse(*componentParams.MemoryLimit)
+	if memoryLimitQuantity.Cmp(requestedMemoryQuantity) < 0 {
+		componentParams.MemoryLimit = ptr.String(requestedMemoryQuantity.String())
+		memoryLimitQuantity = requestedMemoryQuantity
+	}
+}
+
+// SetCPUDefaults sets the CPU consumption related default field values for ComponentParameters.
+func (componentParams *ComponentParameters) SetCPUDefaults(defaultCPURequest, defaultCPULimit string) {
+	if componentParams.CPURequest == nil {
+		componentParams.CPURequest = ptr.String(defaultCPURequest)
+	}
+	if componentParams.CPULimit == nil {
+		componentParams.CPULimit = ptr.String(defaultCPULimit)
 	}
```

Comment:
This is problematic. If I explicitly set a limit (e.g. 500m) but without request. Then this logic would default the request to something like 1500m which is not correct.

I suggest group request/limit together like

```
type ComponentParameters struct {
  CPU *SystemResource
  Memory *SystemResource
  // ...
}

type SystemResource struct {
  Request, Limit string
}
```

If `SystemResource` is specified for CPU, then we don't set its value anymore. We only set defaults if `SystemResource` is nil. That way we know if a request/limit is explicitly set to empty or it's something we need to set default.

Reply:
Thanks, the re-factored version now has a structure with groupings similar to what's suggested above, with slight variation: to make the data contract match the corev1 requests and limits structure.